circuit_maintenance_parser/utils.py

```diff
@@ -1,27 +1,116 @@
+            cls.db_location = pd.read_csv(os.path.join(dirname, "data", "worldcities.csv"))
+            logger.info("Loaded local locations DB.")
+
+    def get_location(self, city: str) -> Tuple[float, float]:
+        """Get location."""
+        try:
+            location_coordinates = self.get_location_from_local_file(city)
+        except ValueError:
+            location_coordinates = self.get_location_from_api(city)
+
+        logger.debug(
+            "Resolved city %s to coordinates: lat %s - lon %s", city, location_coordinates[0], location_coordinates[1],
+        )
+        return location_coordinates
+
+    def get_location_from_local_file(self, city: str) -> Tuple[float, float]:
+        """Get location from Local DB."""
+        city_name = city.split(", ")[0]
+        country = city.split(", ")[-1]
+
+        if self.db_location is not None:
+            # 1. We try to match city name and country
+            location = self.db_location.query(f"city_ascii == '{city_name}' & country == '{country}'")
+            if not location.empty:
+                logger.debug("Used city name and country to resolve %s to %s from local locations DB.", city, location)
+                return (location.lat.values[0], location.lng.values[0])
+
+            # 2. We try to match only city name
+            location = self.db_location.query(f"city_ascii == '{city_name}'")
+            if not location.empty:
+                logger.debug("Used only city name to resolve %s to %s from local locations DB.", city, location)
+                return (location.lat.values[0], location.lng.values[0])
+
+        logger.debug("City %s was not resolvable in the local locations DB.", city)
+        raise ValueError
+
+    @staticmethod
+    @backoff.on_exception(
+        backoff.expo, (GeocoderUnavailable, GeocoderTimedOut, GeocoderServiceError), max_time=10, logger=logger,
+    )
+    def get_location_from_api(city: str) -> Tuple[float, float]:
+        """Get location from API."""
         geolocator = Nominatim(user_agent="circuit_maintenance")
         location = geolocator.geocode(city)  # API call to OpenStreetMap web service
-        timezone = (
-            tzwhere.tzwhere()
-        )  # TODO: Offline loading of timezone location data is quite slow. Look for better alternative
-        return timezone.tzNameAt(location.latitude, location.longitude)
-    except (GeocoderUnavailable, GeocoderTimedOut, GeocoderServiceError):
-        raise ParserError(  # pylint: disable=raise-missing-from
-            "Cannot connect to the remote Geolocator API to determine timezone"
-        )
+        logger.debug("Resolved %s to %s from OpenStreetMap webservice.", city, location)
+        return (location.latitude, location.longitude)
+
+    def city_timezone(self, city: str) -> str:
+        """Get the timezone for a given city.
+
+        Args:
+            city (str): Geographic location name
+        """
+        if self.timezone is not None:
```

Comment:
Is this check needed? Same reason.

Reply:
same as before, maybe it assumes that the method is available for class, no instance and fails the validation